### PR TITLE
Fix null value handling of NOT Predicate

### PR DESCRIPTION
**Pull request by mkleen**
## Summary of the changes / Why this improves CrateDB

This simplifies the Three-Valued-Logic Optimization (3vl) for the `NotPredicate`. 
Instead of explicitly including the functions where 3vl is needed, this makes 3vl the default behaviour and handles the cases where the 3vl can be ignored: 

- No function is involved `NOT x`
- A comparison operator from a reference to a value `NOT x = 1`
- ignore3vl function `ignore3vl( ... )`

Resolves https://github.com/crate/crate/issues/15083

## Checklist

 - [x] Added an entry in the latest `docs/appendices/release-notes/<x.y.0>.rst` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in the latest `docs/appendices/release-notes/<x.y.0>.rst`
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: server/src/main/java/io/crate/expression/predicate/NotPredicate.java [hunk 1]
```diff
@@ -112,50 +110,48 @@ void add(Reference symbol) {
...
-                for (Symbol arg : function.arguments()) {
-                    arg.accept(this, context);
+            } else if (EqOperator.NAME.equals(functionName)) {
+                var a = function.arguments().get(0);
+                var b = function.arguments().get(1);
+                // in the  case x = `foo` or foo = `x` 3vl can be skipped
+                if (a instanceof Reference && b instanceof Literal<?> ||
+                    a instanceof Literal<?> && a instanceof Reference) {
+                    context.skip3ValuedLogic = true;
                 }
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/expression/predicate/NotPredicate.java:149]: Wouldn't this also apply to all other comparison operators?

Comment (mkleen) [reply server/src/main/java/io/crate/expression/predicate/NotPredicate.java:149]: Yes true, that's a good point.

Comment (mfussenegger) [reply server/src/main/java/io/crate/expression/predicate/NotPredicate.java]: Or is there a more general property we could identify that let's us skip 3vl? 

Maybe as part of a Function Feature https://github.com/crate/crate/blob/6f2278fa55d061e576e8f0c4e69ecc340ee4d2b6/server/src/main/java/io/crate/metadata/Scalar.java#L152 ?

E.g. `x = 'foo'` would now get optimized, but `x` = 'fo' || 'o'` wouldn't (unless it's normalized already) ?

I'm a bit worried that we're causing a huge performance regression here for cases that did and could get optimized

Comment (mkleen) [reply server/src/main/java/io/crate/expression/predicate/NotPredicate.java]: So you would define it as a function feature that we can skip 3vl ?

Comment (mfussenegger) [reply server/src/main/java/io/crate/expression/predicate/NotPredicate.java]: No, I don't know what I would do. I'm asking if there's a more general aspect that would let us avoid having to list all exceptions. Or at least identify what property a function needs to have to ignore 3vl in NOT

If we do need to provide some information, then we _could_ use the scalar feature properties to communicate that - if it helps.

The initial assumption was that some functions like `coalesce` can recover a `null` into a non-null value. Only a few functions do that.
Another aspect are arrays - where even if the value isn't null, an element might be. 

E.g. one question is why the `NOT (c0||0)<-1` from the latest issue doesn't work. There's no null-recovery for `c0`

Comment (mkleen) [reply server/src/main/java/io/crate/expression/predicate/NotPredicate.java]: Let me think if there is a better way to handle this.

**Pull request closed without merging**